frontend/packages/console-plugin-sdk/src/withExtensions.tsx

```diff
@@ -0,0 +1,99 @@
+import { RootState } from '@console/internal/redux';
+import { stateToProps } from '@console/internal/reducers/features';
+import { pluginStore } from '@console/internal/plugins';
+import { getGatingFlagNames, isExtensionInUse } from './store';
+import { Extension, ExtensionTypeGuard } from './typings';
+
+/**
+ * React higher-order component (HOC) for consuming Console extensions.
+ *
+ * This is semantically equivalent to `useExtensions` hook with one difference:
+ * this HOC supports retrieving different extension types and providing them to
+ * your component as props.
+ *
+ * Example usage:
+ *
+ * ```ts
+ * import {
+ *   withExtensions,
+ *   NavItem,
+ *   Perspective,
+ *   isNavItem,
+ *   isPerspective,
+ * } from '@console/plugin-sdk';
+ *
+ * const Example = withExtensions<ExampleExtensionProps>({
+ *   navItemExtensions: isNavItem,
+ *   perspectiveExtensions: isPerspective,
+ * })(
+ *   class Example extends React.Component<ExampleOwnProps & ExampleExtensionProps> {
+ *     render() {
+ *       const { navItemExtensions, perspectiveExtensions } = this.props;
+ *       // process extensions and render your component
+ *     }
+ *   },
+ * );
+ *
+ * type ExampleExtensionProps = {
+ *   navItemExtensions: NavItem[];
+ *   perspectiveExtensions: Perspective[];
+ * };
+ * ```
+ *
+ * @param typeGuardMapper Object that maps prop names to extension type guards.
+ * It's basically an object-based analogy to Redux `mapStateToProps` function.
+ */
+export const withExtensions: WithExtensions = (typeGuardMapper) => {
+  const allExtensions = pluginStore.getAllExtensions();
+
+  // 1) Narrow extensions according to type guards
+  const matchedExtensions = _.flatMap(
+    Object.values(typeGuardMapper).map((tg) => allExtensions.filter(tg)),
+  );
+
+  // 2.a) Compute flags relevant for gating matched extensions
+  const gatingFlagNames = getGatingFlagNames(matchedExtensions);
+
+  return connect(
+    (state: RootState) => {
+      // 2.b) Compute flags relevant for gating matched extensions
+      const gatingFlags = stateToProps(gatingFlagNames, state).flags;
+
+      // 3) Gate matched extensions by relevant feature flags
+      const extensionsInUse = matchedExtensions.filter((e) => isExtensionInUse(e, gatingFlags));
+
+      return Object.keys(typeGuardMapper).reduce(
+        (props, propName) => ({
+          ...props,
+          [propName]: extensionsInUse.filter(typeGuardMapper[propName]),
+        }),
+        {},
+      );
+    },
+    null,
+    null,
+    {
+      areStatesEqual: ({ FLAGS: next }, { FLAGS: prev }) =>
+        gatingFlagNames.every((f) => next.get(f) === prev.get(f)),
+    },
+  );
+};
+
+type WithExtensions = <P extends WithExtensionsProps<E>, E extends Extension = Extension>(
+  typeGuardMapper: WithExtensionsTypeGuardMapper<E, P>,
+) => (
+  C: React.ComponentType<P>,
+) => React.ComponentType<Omit<P, keyof WithExtensionsProps<E>>> & {
+  WrappedComponent: React.ComponentType<P>;
+};
```

Comment:
Much prefer the object param to type the incoming extension props!

Using this HOC typescript won't let me provide own props to the wrapped component.
Try something like this:
```
type WithExtensions = <
  E extends Extension,
  P extends WithExtensionsProps<E>,
  T extends WithExtensionsTypeGuardMapper<E, P>
>(
  typeGuardMapper: T,
) => <OP extends P>(
  C: React.ComponentType<OP>,
) => React.ComponentType<Omit<OP, keyof T>> & {
  WrappedComponent: React.ComponentType<OP>;
};
```